### PR TITLE
Framework: Refactor away from `_.first()`

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

Lodash's `first()` is only several times the Calypso codebase, and the usage is pretty simple, so this PR replaces it with `[ 0 ]` and optional chaining where necessary. 

I'm consciously not refactoring other neighboring or related instances that can be touched (`head()` or `last()`) - these will be addressed separately, as I'm trying to keep changes small, contained, and easier to test.

If you want to find out why we're moving away from Lodash, see https://github.com/Automattic/wp-calypso/pull/50368 and p4TIVU-9Bf-p2.

#### Testing instructions

* Verify all tests pass.
* Verify `/home/:site` still works well - specifically, inline help.
* Verify domain search (`/domains/add/:site`) still works well.
* Verify Line Chart still works well (`/devdocs/design/line-chart`)
* Verify notification settings still works well (`/me/notifications`)
* Verify revision comparison still in post editor still works well.

### CODE REVIEW

File: client/blocks/inline-help/contextual-help.js [hunk 1]
```diff
@@ -1485,8 +1485,8 @@ export function getContextResults( section ) {
 
 	// make sure editorially to show at most one tour and one video at once
 	// `first` is a safe-guard in case that fails
-	const video = first( get( videosForSection, section ) );
-	const tour = first( get( toursForSection, section ) );
+	const video = get( videosForSection, section )?.shift();
```
Comment (flootr) [inline client/blocks/inline-help/contextual-help.js]: Doesn't `Array.shift` mutate the array? That'd be different to `.first()` which gets the first element, but let's the array in place. Is that something we should consider? Or does it matter for those use cases at all?

Comment (sarayourfriend) [reply client/blocks/inline-help/contextual-help.js]:
```
> arr = [1,2,3]
[ 1, 2, 3 ]
> arr.shift()
1
> arr
[ 2, 3 ]
```

It does mutate the array, I don't think it's a safe way to do this.

What's wrong with using `[0]` for all of these cases?

Comment (tyxla) [reply client/blocks/inline-help/contextual-help.js]: I agree it does mutate the array, but as long as we don't use the array anymore, it should be fine, no?

Comment (jsnajdr) [reply client/blocks/inline-help/contextual-help.js]: `videosForSection[ section ]?.[ 0 ]` looks like a perfect replacement to me 🙂

Comment (sarayourfriend) [reply client/blocks/inline-help/contextual-help.js]: `get( videosForSection, section )` retrieves the reference, there's no guarantee that the array isn't being used somewhere else.

Comment (jsnajdr) [reply client/blocks/inline-help/contextual-help.js]: > but as long as we don't use the array anymore

The `videosForSection` array is a global variable. Each call to `getContextResult` (it's called in a Redux selector) will eat one item from the array, like Pacman 🙂

Comment (tyxla) [reply client/blocks/inline-help/contextual-help.js]: I'm glad y'all are paying attention to the detail 👏 😉  Going to fix this tomorrow.

File: client/state/stats/chart-tabs/reducer.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ const countsReducer = ( state = [], action ) => {
 			// See https://github.com/Automattic/wp-calypso/pull/41441#discussion_r415918092
 			if (
 				action.data.length !== state.length ||
-				! isEqual( first( action.data ).period, first( state ).period )
+				! isEqual( action.data[ 0 ].period, state[ 0 ].period )
```
Comment (jsnajdr) [inline client/state/stats/chart-tabs/reducer.js:33]: I know this code has been here before, but `data` items are themselves arrays. There is not `.period` field. Response from the `/sites/:site/stats/visits` endpoint looks like this:

Screenshot 2021-04-20 at 14 14 52

Note how the `fields` array specifies what the `data[ x ][ y ]` fields contain.

Comment (tyxla) [reply client/state/stats/chart-tabs/reducer.js:33]: Yes, but the array of fields is transformed to an object at the data layer level - see https://github.com/Automattic/wp-calypso/blob/remove/lodash-first/client/state/data-layer/wpcom/sites/stats/visits/from-api.js#L8 and https://github.com/Automattic/wp-calypso/blob/remove/lodash-first/client/state/stats/lists/utils.js#L239-L269

Comment (jsnajdr) [reply client/state/stats/chart-tabs/reducer.js:33]: I see, mystery solved 🙂 

Still, `.period` is always a string, never an object. `isEqual` should not be necessary. Merely comparing them with `===` should suffice.

**Review by jsnajdr — APPROVED** [review]
Looks good except the redundant `isEqual` in the `STATS_CHART_COUNTS_RECEIVE`. That's not even a bug, just a pre-existing inefficiency (and Lodash usage 🙂)

File: client/state/stats/chart-tabs/reducer.js (hunk 2 above)
Comment (tyxla) [reply client/state/stats/chart-tabs/reducer.js:33]: Agreed, but I prefer skipping it for this PR - there's another `isEqual` usage there that compares objects, and all this isn't directly related to this PR.

**Pull request merged**